airflow/contrib/hooks/bigquery_hook.py

```diff
@@ -656,6 +668,21 @@ def run_query(self,
                 configuration['query'][
                     'schemaUpdateOptions'] = schema_update_options
 
+        if not api_resource_configs:
+            api_resource_configs = self.api_resource_configs
+
+        if 'configuration' in api_resource_configs and isinstance(
```

Comment:
I'm having a hard time understanding this logic with all the nested loops and type checking.

I think it'd be easier if we started off by setting `configuration = copy.deepcopy(api_resource_configs)` just before https://github.com/apache/incubator-airflow/blob/e933ee8e52e0449d88bac0c83f6ee06729988343/airflow/contrib/hooks/bigquery_hook.py#L602

We should warn anywhere where we set a `key` by one of the other arguments but it was already present in `api_resource_configs`.

Reply:
Do you mean first do a copy of api_config in configuration and then add other params to configuration if they do not exist? Am I right understand?

Reply:
check pls

Reply:
Yes. Since `api_resource_configs` basically _is_ `configuration`, a straight `deepcopy` should work as a first step.